lib/linguist/heuristics.rb

```diff
@@ -465,5 +465,13 @@ def call(data)
         Language["Scilab"]
       end
     end
+
+    disambiguate ".tsx" do |data|
+      if /^(import.+from|.+=\s*require)[\s\(]+['"]react['"]/.match(data)
+        Language["TypeScript"]
+      else
+        Language["XML"]
```

Comment:
This `else` is too dangerous: any file that doesn't match the first regular expression will be classified as XML :/
Couldn't we match on `<?xml version` instead?